test/EtomicSwap.js

```diff
@@ -1,16 +1,22 @@
-const Swap = artifacts.require('EtomicSwap');
-const Token = artifacts.require('Token');
-const Erc721Token = artifacts.require('Erc721Token');
-const Erc1155Token = artifacts.require('Erc1155Token');
+const {
+    expect
+} = require("chai");
+const {
+    ethers
+} = require("hardhat");
 const crypto = require('crypto');
 const RIPEMD160 = require('ripemd160');
-
-const EVMThrow = 'VM Exception while processing transaction';
+const {
+    AbiCoder
+} = require("ethers");
 
 require('chai')
     .use(require('chai-as-promised'))
     .should();
 
+const REVERT_WITHOUT_REASON = 'Transaction reverted without a reason';
```

Comment:
Hmm I got what you mean. Good note,
Just small question/note.
For example  in `erc20Payment` method currently we have this require
```solidity
require(
            receiver != address(0) &&
            amount > 0 &&
            payments[id].state == PaymentState.Uninitialized
        );
```
We can split the conditions and add specific error messages for each:
```solidity
require(receiver != address(0), "Receiver cannot be zero address.");
require(amount > 0, "Amount must be greater than 0.");
require(payments[id].state == PaymentState.Uninitialized, "Payment already initialized.");
```
I want to note  that having multiple `require()` statements instead of combining conditions into a single `require()` statement can have a slight impact on gas usage, potentially increasing it. Each `require()` statement consumes a bit of gas to evaluate its condition.
But such strategy increases code readability and maintainability.

@artemii235 are you agree to use multiple statements?
I dont mind to keep using one require statement even if it saves some gas usage just slightly (but still).

Reply:
UPD: I will just push my solution and we will discuss it

Reply:
What is gas usage difference between one and multiple `require`?

Reply:
> What is gas usage difference between one and multiple `require`?

Interesting. Did tests and some research. 
Shortly: yes, having more `require` can increase gas usage. 
However, when we are talking about logical `AND` operation in one `require`, this also increases gas usage, while splitting such checks in separate `require` statements in opposite is more gas effective. 
But [this will cost more to deploy, but can save gas when the function is run](https://cryptoguide.dev/posts/solidity-gas-optimizations-guide/#use-separate-checks-in-require()-to-save-gas). 

Here what I had in console log during several test iterations:
Used **"should allow receiver to spend ERC721 payment by revealing a secret"** test.

One `require` with multiple logical  `&&` operations
Note: Where you see **96113** gas used in  `ERC721 safeTransferFrom` iterations, here I significantly reduced error message in `onERC721Received` EtomicSwap method. Then return error message to the current state and have got **96125** gas usage
```
Gas Used ERC721 safeTransferFrom: 96113
Gas Used receiverSpendErc721: 74823

Gas Used ERC721 safeTransferFrom: 96113
Gas Used receiverSpendErc721: 74835

Gas Used ERC721 safeTransferFrom: 96125
Gas Used receiverSpendErc721: 74847
```

Multiple `require` with single check 
```
Gas Used ERC721 safeTransferFrom: 96093
Gas Used receiverSpendErc721: 74827

Gas Used ERC721 safeTransferFrom: 96093
Gas Used receiverSpendErc721: 74839

Gas Used ERC721 safeTransferFrom: 96093
Gas Used receiverSpendErc721: 74839
```

@artemii235 to sum up. yeah, we should split our multiple checks into separate `require` statements. Will do it with more specific error messages.

Reply:
Great, thanks for the detailed research! :slightly_smiling_face: